### PR TITLE
Workspace entity tables

### PR DESCRIPTION
**Pull request by dvoet**
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Merge to develop 
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: Check configuration files in Jenkins in case they need changes
- [ ] **Submitter**: Verify swagger UI on dev server still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala [hunk 1]
```diff
@@ -24,12 +29,114 @@ trait AttributeComponent {
...
     def listIndex = column[Option[Int]]("list_index")
 
-    def * = (id, name, valueString, valueNumber, valueDate, valueBoolean, listIndex) <> (AttributeRecord.tupled, AttributeRecord.unapply)
+    def * = (id, name, valueString, valueNumber, valueBoolean, valueEntityRef, listIndex) <> (AttributeRecord.tupled, AttributeRecord.unapply)
+
+    def workspace = foreignKey("FK_ATTRIBUTE_ENTITY_REF", valueEntityRef, entityQuery)(_.id.?)
   }
 
-  protected val attributeQuery = TableQuery[AttributeTable]
+  protected object attributeQuery extends TableQuery(new AttributeTable(_)) {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala:41]: I like this structure of extending TableQuery (as mentioned [here](http://slick.typesafe.com/doc/3.1.1/schemas.html#table-query) ). This gives us a namespace for our functions so our data access component won't have 10 overloaded save functions.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala [hunk 2]
```diff
@@ -10,12 +15,12 @@ case class AttributeRecord(id: Long,
                            name: String,
                            valueString: Option[String],
                            valueNumber: Option[Double],
-                           valueDate: Option[Timestamp],
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala:13]: Why remove date?

Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala:13]: because there is no AttributeDate case class which means we don't actually support it now.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala [hunk 3]
```diff
@@ -24,12 +29,114 @@ trait AttributeComponent {
...
+        case ref: AttributeEntityReference => Seq(insertAttributeRef(name, workspaceId, ref))
+        case AttributeEmptyList => Seq(insertAttributeValue(name, AttributeNull, Option(-1))) // storing empty list as an element with index -1
+      }
+    }
+
+    private def insertAttributeRef(name: String, workspaceId: UUID, ref: AttributeEntityReference, listIndex: Option[Int] = None) = {
+      findEntityByName(workspaceId, ref.entityType, ref.entityName).result.flatMap {
+        case Seq() => throw new RawlsException(s"$ref not found in workspace $workspaceId")
+        case Seq(entityRecord) => (attributeQuery returning attributeQuery.map(_.id)) += marshalAttributeEntityReference(name, listIndex, entityRecord)
+        case _ => throw new RawlsException(s"more than one $ref found in workspace $workspaceId")
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala]: Can this ever happen?  It would violate EntityTable's unique key.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 4]
```diff
@@ -57,12 +62,128 @@ trait WorkspaceComponent {
...
+
+    private def insertWorkspaceAttributeMapping(workspaceRecord: WorkspaceRecord, attrId: Long): FixedSqlAction[Int, driver.api.NoStream, Write] = {
+      workspaceAttributeQuery += WorkspaceAttributeRecord(workspaceRecord.id, attrId)
+    }
+
+    private def deleteWorkspaceAttributeMappings(attributeRecords: Seq[AttributeRecord]): FixedSqlAction[Int, driver.api.NoStream, Write] = {
+      workspaceAttributeQuery.filter(_.attributeId inSet (attributeRecords.map(_.id))).delete
+    }
+
+    private def findByNameQuery(workspaceName: WorkspaceName): driver.api.Query[WorkspaceTable, WorkspaceRecord, Seq] = {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: What do you think about adding a `type` definition for easier reading?  I tried adding this inside the object and it worked:
`private type WorkspaceQueryType = driver.api.Query[WorkspaceTable, WorkspaceRecord, Seq]`

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 5]
```diff
@@ -57,12 +62,128 @@ trait WorkspaceComponent {
...
+
+    def findByName(workspaceName: WorkspaceName): DBIOAction[Option[Workspace], NoStream, Effect.Read] = {
+      loadWorkspace(findByNameQuery(workspaceName))
+    }
+
+    def findById(workspaceId: String): DBIOAction[Option[Workspace], NoStream, Effect.Read] = {
+      loadWorkspace(findByIdQuery(UUID.fromString(workspaceId)))
+    }
+
+    def delete(workspaceName: WorkspaceName): DBIOAction[Boolean, NoStream, Effect.Read with Effect.Write] = {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:111]: I have custom type ideas for `Action`s as well, which will appear in my next PR.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala (hunk 3 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala]: while you and I know this, scala does not, I thought the catch all would make scala happier

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 6]
```diff
@@ -57,12 +62,128 @@ trait WorkspaceComponent {
...
     def * = (workspaceId, groupName, accessLevel) <> (WorkspaceAccessRecord.tupled, WorkspaceAccessRecord.unapply)
   }
 
-  protected val workspaceQuery = TableQuery[WorkspaceTable]
   protected val workspaceAttributeQuery = TableQuery[WorkspaceAttributeTable]
   protected val workspaceAccessQuery = TableQuery[WorkspaceAccessTable]
+
+  object workspaceQuery extends TableQuery(new WorkspaceTable(_)) {
+    def workspaceAccessGroups(workspaceId: UUID) = for {
+      workspaceRec <- workspaceQuery if workspaceRec.id === workspaceId
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: Why is this query needed?  The workspaceAccessQuery call alone looks sufficient.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 4 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: how does having a type called WorkspaceQueryType make reading any easier? Seeing that in the code would lead me to wonder what that actually is then I have to go somewhere else to find it.

Personally, I don't usually like type defs because they hide the information I am looking for. 

But if having it makes it easier for you to follow I will add it.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 5 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:111]: good, cause this stuff sucks

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 7]
```diff
@@ -57,12 +62,128 @@ trait WorkspaceComponent {
...
+
+    def findById(workspaceId: String): DBIOAction[Option[Workspace], NoStream, Effect.Read] = {
+      loadWorkspace(findByIdQuery(UUID.fromString(workspaceId)))
+    }
+
+    def delete(workspaceName: WorkspaceName): DBIOAction[Boolean, NoStream, Effect.Read with Effect.Write] = {
+      findByNameQuery(workspaceName).result.flatMap {
+        case Seq() => DBIO.successful(false)
+        case Seq(workspaceRecord) =>
+          workspaceAttributes(workspaceRecord.id).result.flatMap(recs => DBIO.seq(deleteWorkspaceAttributes(recs):_*)) flatMap { _ =>
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:115]: Could you discuss the use of `andThen` in save vs. `flatMap` here?  I'm not understanding the difference.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 8]
```diff
@@ -57,12 +62,128 @@ trait WorkspaceComponent {
...
+            findByIdQuery(workspaceRecord.id).delete
+          } map { count =>
+            count > 0
+          }
+        case tooMany => throw new RawlsException(s"more than one record found for workspace query: $tooMany")
+      }
+    }
+
+    private def workspaceAttributes(workspaceId: UUID) = for {
+      workspaceRec <- workspaceQuery if workspaceRec.id === workspaceId
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: Same as above: why is this query needed?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 9]
```diff
@@ -57,12 +62,128 @@ trait WorkspaceComponent {
...
+
+    private def findByNameQuery(workspaceName: WorkspaceName): driver.api.Query[WorkspaceTable, WorkspaceRecord, Seq] = {
+      workspaceQuery.filter(rec => rec.namespace === workspaceName.namespace && rec.name === workspaceName.name)
+    }
+
+    private def findByIdQuery(workspaceId: UUID): driver.api.Query[WorkspaceTable, WorkspaceRecord, Seq] = {
+      workspaceQuery.filter(_.id === workspaceId)
+    }
+
+    private def loadWorkspace(lookup: driver.api.Query[WorkspaceTable, WorkspaceRecord, Seq]): DBIOAction[Option[Workspace], NoStream, Read with Read] = {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: Read with Read ?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 7 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:115]: andThen is merely ordering where the second thing does not care about the results of the first thing. flatMap (and map) also does the ordering but passes the result of the first part to the second part.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 9 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: most of these private functions are the results of intellij refactorings (extract method) which seem to put in some wacky things sometimes

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala (hunk 1 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala:41]: :+1:

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala (hunk 3 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala]: ok

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala (hunk 2 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala:13]: that is a very good reason

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 4 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: This is a case where I would prefer it to be hidden, as `Query[WorkspaceTable, WorkspaceRecord, Seq]` doesn't mean a lot to me, but "the standard query type for this table" does.  Differences between similar types will be quicker to discern, as well.

Happy to hear additional opinions on this, though.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 7 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:115]: Sorry - more specifically I mean the `flatMap { _ =>` case where the passed-in value is not used.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 9 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: ah.  still want it fixed though :)

File: src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponentSpec.scala [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+/**
+ * Created by dvoet on 2/9/16.
+ */
+class AttributeComponentSpec extends TestDriverComponent with AttributeComponent {
+  import driver.api._
+
+  "AttributeComponent" should "insert string attribute" in {
+    val testAttribute = AttributeString("test")
+    val insertedIds = attributeQuery.insertAttributeRecords("test", testAttribute, UUID.randomUUID()).map(x => runAndWait(x))
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponentSpec.scala:17]: These can be `map(runAndWait(_))` but I don't care enough to ask for this change.  Do it if you like.

Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponentSpec.scala:17]: Make a note that workspaceID is irrelevant to Value Attributes, or give it a name like arbitraryWorkspace.

I considered the possibility that it should be an Option, but it will only be called by services that will have a real Workspace ID, so there's no need to support calling with None, and it's good as-is.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala [hunk 11]
```diff
@@ -24,12 +29,114 @@ trait AttributeComponent {
...
+    def workspace = foreignKey("FK_ATTRIBUTE_ENTITY_REF", valueEntityRef, entityQuery)(_.id.?)
   }
 
-  protected val attributeQuery = TableQuery[AttributeTable]
+  protected object attributeQuery extends TableQuery(new AttributeTable(_)) {
+    def insertAttributeRecords(name: String, attribute: Attribute, workspaceId: UUID) = {
+      attribute match {
+        case AttributeEntityReferenceList(refs) => refs.zipWithIndex.map { case (ref, index) => insertAttributeRef(name, workspaceId, ref, Option(index))}
+        case AttributeValueList(values) =>
+          assertConsistentListMembers(values)
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala]: Do we want a similar `assertConsistentEntityType` for `case AttributeEntityReferenceList`?  Is it OK to mix entity types in a reference list?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala [hunk 12]
```diff
@@ -24,12 +29,114 @@ trait AttributeComponent {
...
+          values.zipWithIndex.map { case (value, index) => insertAttributeValue(name, value, Option(index))}
+        case value: AttributeValue => Seq(insertAttributeValue(name, value))
+        case ref: AttributeEntityReference => Seq(insertAttributeRef(name, workspaceId, ref))
+        case AttributeEmptyList => Seq(insertAttributeValue(name, AttributeNull, Option(-1))) // storing empty list as an element with index -1
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala]: I don't like this but it works and I don't have a better implementation.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 13]
```diff
@@ -57,12 +62,128 @@ trait WorkspaceComponent {
...
 
     def * = (workspaceId, groupName, accessLevel) <> (WorkspaceAccessRecord.tupled, WorkspaceAccessRecord.unapply)
   }
 
-  protected val workspaceQuery = TableQuery[WorkspaceTable]
   protected val workspaceAttributeQuery = TableQuery[WorkspaceAttributeTable]
   protected val workspaceAccessQuery = TableQuery[WorkspaceAccessTable]
+
+  object workspaceQuery extends TableQuery(new WorkspaceTable(_)) {
+    def workspaceAccessGroups(workspaceId: UUID) = for {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: Should this be public?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 7 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:115]: meh... looks like they are equivalent

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 14]
```diff
@@ -57,12 +62,128 @@ trait WorkspaceComponent {
...
+      workspaceRec <- workspaceQuery if workspaceRec.id === workspaceId
+      workspaceAccessRec <- workspaceAccessQuery if workspaceRec.id === workspaceAccessRec.workspaceId
+    } yield (workspaceAccessRec)
+
+    def save(workspace: Workspace): DBIOAction[Workspace, NoStream, Effect.Read with Effect.Write] = {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: I don't see any enforcement of uniqueness for namespace+name.  Add a test that adds a second Workspace with a duplicate namespace+name.  findByName and delete will throw RawlsExceptions on retrieval of the duplicate entries, so we should stop them from appearing in the first place.

File: src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponentSpec.scala [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+    }
+    val attributeRecs = runAndWait(attributeQuery.filter(_.id inSet inserts).result)
+    assertResult(3) { attributeRecs.size }
+
+    assertResult(3) { runAndWait(attributeQuery.deleteAttributeRecords(attributeRecs)) }
+
+    assertResult(0) { runAndWait(attributeQuery.filter(_.id inSet inserts).result).size }
+  }
+
+  it should "unmarshall attribute records" in {
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponentSpec.scala:154]: Add a test for the case where some, but not all, members of a list have None for the index

**Comment by jmthibault79** [comment]
First round of review done.

**Comment by jmthibault79** [comment]
note: Entity DAO and Spec are still TODO and will be covered in a different PR

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 13 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: removing it anyway

**Comment by dvoet** [comment]
@jmthibault79 review comments addressed.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 16]
```diff
@@ -36,6 +41,8 @@ trait WorkspaceComponent {
     def createdBy = column[String]("created_by")
     def isLocked = column[Boolean]("is_locked")
 
+    def uniqueNamespaceName = index("IDX_WS_UNIQE_NAMESPACE_NAME", (namespace, name), unique = true)
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: Typo

**Comment by jmthibault79** [comment]
:+1: after a typo fix (WC line 44)
